effectful/ops/semantics.py

```diff
@@ -228,15 +243,53 @@ def evaluate[T](expr: Expr[T], *, intp: Interpretation | None = None) -> Expr[T]
     6
 
     """
+    import inspect
+
     from effectful.internals.runtime import get_interpretation, interpreter
 
     if intp is not None:
         return interpreter(intp)(evaluate)(expr)
 
     if isinstance(expr, Term):
-        args = tuple(evaluate(arg) for arg in expr.args)
-        kwargs = {k: evaluate(v) for k, v in expr.kwargs.items()}
-        return expr.op(*args, **kwargs)
+        # Check if we need async evaluation
+        args_evaluated = [evaluate(arg) for arg in expr.args]
+        kwargs_evaluated = {k: evaluate(v) for k, v in expr.kwargs.items()}
+
+        # Check if any evaluated args/kwargs are coroutines
+        has_coro = any(inspect.iscoroutine(a) for a in args_evaluated) or any(
```

Comment:
Can we push this logic into an `if isinstance(expr, Coroutine)` branch in `evaluate`:
```py
...
if isinstance(expr, Term):
    args = tuple(evaluate(arg) for arg in expr.args)
    kwargs = {k: evaluate(v) for k, v in expr.kwargs.items()}
    result = expr.op(*args, **kwargs)
    if isinstance(result, Coroutine):
        return await result
    else:
        return result
elif isinstance(expr, Coroutine):
    return await expr
...
```